opm/simulators/flow/ReservoirCoupling.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+  Copyright 2024 Equinor AS
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+
+#include <opm/simulators/flow/ReservoirCoupling.hpp>
+#include <opm/simulators/utils/ParallelCommunication.hpp>
+#include <opm/common/OpmLog/OpmLog.hpp>
+
+#include <fmt/format.h>
+
+namespace Opm {
+namespace ReservoirCoupling {
+
+void custom_error_handler_(MPI_Comm* comm, int* err, const std::string &msg)
+{
+    // It can be useful to have a custom error handler for debugging purposes.
+    // If not, MPI will use the default error handler which aborts the program, and
+    // it can be difficult to determine exactly where the error occurred. With a custom
+    // error handler you can at least set a breakpoint here to get a backtrace.
+    int rank;
+    MPI_Comm_rank(*comm, &rank);
+    char err_string[MPI_MAX_ERROR_STRING];
+    int len;
+    MPI_Error_string(*err, err_string, &len);
+    const std::string error_msg = fmt::format(
+        "Reservoir coupling MPI error handler {} rank {}: {}", msg, rank, err_string);
+    // NOTE: The output to terminal vie stderr or stdout has been redirected to files for
+    //   the slaves, see Main.cpp. So the following output will not be visible in the terminal
+    //   if we are called from a slave process.
+    // std::cerr << error_msg << std::endl;
+    OpmLog::error(error_msg);  // Output to log file, also shows the message on stdout for master.
+    MPI_Abort(*comm, *err);
+}
+
+void custom_error_handler_slave_(MPI_Comm* comm, int* err, ...)
+{
+    custom_error_handler_(comm, err, "slave");
+}
+
+void custom_error_handler_master_(MPI_Comm* comm, int* err, ...)
+{
+    custom_error_handler_(comm, err, "master");
+}
+
+void setErrhandler(MPI_Comm comm, bool is_master)
+{
+    MPI_Errhandler errhandler;
+    // NOTE: Lambdas with captures cannot be used as C function pointers, also
+    //   converting lambdas that use ellipsis "..." as last argument to a C function pointer
+    //   is not currently possible, so we need to use static functions instead.
+    if (is_master) {
+        MPI_Comm_create_errhandler(custom_error_handler_master_, &errhandler);
+    }
+    else {
+        MPI_Comm_create_errhandler(custom_error_handler_slave_, &errhandler);
+    }
+    MPI_Comm_set_errhandler(comm, errhandler);
```

Comment:
>One would this directly in main or at least way up in the call stack.

Note that these error handlers are set only for the slave-master communicators. So the default error handler still applies to the MPI_COMM_WORLD communicator set up in main. Currently, the slave-master communicators are first defined in `SimulatorFullyImplicitBlackoil.hpp` in init() or run(). And therefore it is currently not convenient to set error handlers for these in `Main.hpp`. 

Can you elaborate on why it is better to use the default error handler for these communicators? I see two advantages with using custom error handlers: Debugging (e.g. setting breakpoints in the handler and getting backtrace), and opportunity to provide more context in the error message.

Reply:
> I doubt that this works as errhandler is a local variable.

As I understand, `errhandler` is just an MPI handle (an integer) so MPI will create its internal errhandler structure corrsponding to that handle when we call `MPI_Comm_create_errhandler()` + `MPI_Comm_set_errhandler(comm, errhandler)`. After that we can simply call `MPI_Errhandler_free()` on that handle. The handle itself (the local `errhandler` variable) is just a copy of an integer and can safely be thrown away as the function `setErrhandler()` returns. I tried to clarify this in the latest commit.

Reply:
To me these functions are for memory management that needs to be done in your code. The handle is basically a pointer to an internally allocated object.

You do have point, though. There is no lifetime issue as this is just handle and the err handler is allocated internally and lives until `MPI_Errhandler_free` is called on it. As this is never called we just have a memory leak. Not nice, but also not fatal.

But if you call `MPI_Errhandler_free`  in this function (as you proposed) then you will have created a memory problem inside of MPI.

The issue that I see with this is code is that I cannot simply overwrite the error handler anymore in Main.cpp or somewhere else up the stack. You unconditionally reset it here. Hence I would to change it here, too.

Reply:
> But if you call MPI_Errhandler_free in this function (as you proposed) then you will have created a memory problem inside of MPI.

After checking the source code of open MPI, I would claim that this is not memory problem inside MPI. Refer to 
[comm_set_errhandler.c](https://github.com/open-mpi/ompi/blob/main/ompi/mpi/c/comm_set_errhandler.c) and [errhandler_free.c](https://github.com/open-mpi/ompi/blob/main/ompi/mpi/c/errhandler_free.c). It shows that the communicator also claims a reference to the errhandler. So it seems correct to call `MPI_Errhandler_free()` immediately. This will release the memory of the errhandler when the communicator is deallocated.